File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {

    // 초기 조건
    const wordSet = new Set(words);
    if (!wordSet.has(target)) {
        return 0;
    }
    
    const n = words.length;
    const m = words[0].length;
    const visited = Array(n).fill(false);
    
    let answer = 0;
    function dfs(current, depth) {
        if (current === target) {
            answer = depth;
            return;
        }
        
        for (let i = 0 ; i < n ; i++) {
            const convertable = getDuplicatedCount(current, words[i]) === m - 1;

            if (!visited[i] && convertable) {
                visited[i] = true;
                dfs(words[i], depth+1);
                visited[i] = false;
            }
        }
    }
    
    dfs(begin, 0);
    return answer;
}

function getDuplicatedCount(a, b) {
    const n = a.length;
    let pointer = 0;
    let count = 0;
    
    while (pointer < n) {
        if (a[pointer] === b[pointer]) {
            count += 1;
        }
        pointer += 1;
    }
    
    return count;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    
    visited = set()
    queue = deque([(begin, 0)])
    
    while queue:
        cur, cnt = queue.popleft()
        
        if cur == target:
            return cnt
        
        for i in range(len(words)):
            if words[i] not in visited and is_convertable(cur, words[i]):
                visited.add(words[i])
                queue.append((words[i], cnt + 1))

    return 0
    
    visited = [False for _ in range(len(words))]
    
    

def is_convertable(word_a, word_b):
    n = len(word_a)
    
    count = 0
    
    for i in range(n):
        
        if word_a[i] != word_b[i]:
            count += 1
            
        if count >= 2:
            return False
    
    if count == 0:
        return False
    
    return True